[doc] Add DataPipeline + Callbacks + Registry (#207)
PyTorchLightning/lightning-flash: docs/source/custom_task.rst

Tutorial: Creating a Custom Task
================================

In this tutorial we will go over the process of creating a custom :class:`~flash.core.model.Task`,
along with a custom :class:`~flash.data.data_module.DataModule`.

The tutorial objective is to create a ``RegressionTask`` to learn to predict if someone has ``diabetes`` or not.

> The diabetes  (stored as a numpy dataset).

Reviewer: Is this about the dataset? not a complete sentance.

Author: Done !

In this tutorial we will go over the process of creating a custom task,
along with a custom data module.

1 . Imports
-----------

.. testcode:: python

    from typing import Any, List, Tuple

    import numpy as np
    import torch
    from pytorch_lightning import seed_everything
    from sklearn import datasets
    from sklearn.model_selection import train_test_split
    from torch import nn

    import flash
    from flash.data.auto_dataset import AutoDataset
    from flash.data.process import Postprocess, Preprocess

    seed_everything(42)

> 2 . The Task: Linear regression

Reviewer 2: I think data should go before task.

Reviewer 2: Steps to create a task

1. DataModule
2. PreProcess
3. Post Process
4. Registry 
5. Task
6. Example
7. Test (This is important since people need to know how to test to merge)

Author: Done ! Feel free to re-organize this when you have time.

-------------------------------

> Here we create a basic linear regression task by subclassing

Reviewer 2: Linear regression is a method. The task should be Forecasting or something.

Author: Renamed `RegressionTask`

testing) or override ``training_step``, ``validation_step``, and
``test_step`` individually. These methods behave identically to PyTorch
Lightning’s
`methods <https://pytorch-lightning.readthedocs.io/en/latest/common/lightning_module.html#methods>`__.

3. The Data
-------------

For a task you will likely need a specific way of loading data.

It is recommended to create a :class:`~flash.data.process.Preprocess` object.

> The :class:`~flash.data.process.Preprocess` contains all the preprocessing logic and works similarly to ``Callback``.

Reviewer 3: ```suggestion
The :class:`~flash.data.process.Preprocess` object contains all of the preprocessing logic and works similarly to a ``Callback``.
```

The user has to override hooks with their processing logic.

.. note::
    As new concepts are being introduced, we strongly encourage the reader to click on :class:`~flash.data.process.Preprocess`

>     before going further in the tutorial.

Reviewer 4: ```suggestion
    before going further with the tutorial.
```

Author: Done !

> The user would have to implement a :class:`~flash.data.data_module.DataModule` as a way to perform data checks and instantiate the preprocess.

Reviewer 4: data checks?

Author: removed !

.. note::

>    Philosophically, the :class:`~flash.data.process.Preprocess` belongs with the :class:`~flash.data.data_module.DataModule`

Reviewer 4: Love the Philosophically haha <3

   and the :class:`~flash.data.process.Postprocess` with the :class:`~flash.core.model.Task`.

3.a The DataModule API
----------------------

First, let's design the user-facing API. The ``NumpyDataModule`` will provide a ``from_xy_dataset`` helper ``classmethod``.

Example::

    x, y = ...
    preprocess_cls = ...
    datamodule = NumpyDataModule.from_xy_dataset(x, y, preprocess_cls)

> Here are the `NumpyDataModule`` implementation:

Reviewer 3: ```suggestion
Here is the `NumpyDataModule`` implementation:
```

Author: Done !

Example::

    from flash import DataModule
    from flash.data.process import Preprocess

    class NumpyDataModule(DataModule):

        @classmethod

>         def from_xy_dataset(cls, x: ND, y: ND, preprocess_cls: Preprocess = NumpyPreprocess, batch_size: int = 64, num_workers: int = 0):

Reviewer 4: What is `ND` typehint?

Reviewer 4: Edit: I saw the reference below

Author: ND = np.ndarray

            # we don't actually need to override this method for this example
            return self.model(x)

.. note::

>    Lightning Flash provides an API to register models within a store.

Reviewer: Not clear what you mean by store?

The Data
--------

For a task you will likely need a specific way of loading data.

Firstly, it is recommended to create a :class:`~flash.data.process.Preprocess` object.
The :class:`~flash.data.process.Preprocess` contains all the processing logic and are similar to ``Callback``.

> The user would to override hooks with their processing logic.

Author: Done !

Lightning’s

> `methods <https://pytorch-lightning.readthedocs.io/en/latest/lightning_module.html#methods>`__.

Reviewer 5: 404 for this link. Can you fix it?

Author: Great catch ! Updated.

> 3 . The Data

Reviewer 2: Data should go before Tastk

Author: Done !

-------------

For a task you will likely need a specific way of loading data.

It is recommended to create a :class:`~flash.data.process.Preprocess` object.

> The :class:`~flash.data.process.Preprocess` contains all the processing logic and are similar to ``Callback``.

Reviewer 5: ```suggestion
The :class:`~flash.data.process.Preprocess` contains all the preprocessing logic and works similarly to ``Callback``.
```

Author: Done !

The user would override hooks with their processing logic.

.. note::
    As new concepts are being introduced, we strongly encourage the reader to click on :class:`~flash.data.process.Preprocess`
    before going further in the tutorial.

The user would have to implement a :class:`~flash.data.data_module.DataModule` as a way to perform data checks and instantiate the preprocess.

.. note::

   Philosophically, the :class:`~flash.data.process.Preprocess` belongs with the :class:`~flash.data.data_module.DataModule`
   and the :class:`~flash.data.process.Postprocess` with the :class:`~flash.Task`.

3.a The DataModule API
----------------------

First, let's design the user-facing API. The ``NumpyDataModule`` will provide a ``from_xy_dataset`` helper ``classmethod``.

Example::

    x, y = ...
    preprocess_cls = ...
    datamodule = NumpyDataModule.from_xy_dataset(x, y, preprocess_cls)

> Here are the

Author: I just pushed them.

    of connecting the provided :class:`~flash.data.process.Preprocess` with the :class:`~flash.data.data_module.DataModule`.
    Make sure to instantiate your :class:`~flash.data.data_module.DataModule` with this helper if you rely on :class:`~flash.data.process.Preprocess`
    objects.

3.b The Preprocess API
----------------------

Example::

    import torch
    from torch import Tensor
    import numpy as np

    ND = np.ndarray

    class NumpyPreprocess(Preprocess):

        def load_data(self, data: Tuple[ND, ND], dataset: AutoDataset) -> List[Tuple[ND, float]]:
            if self.training:
                dataset.num_inputs = data[0].shape[1]
            return [(x, y) for x, y in zip(*data)]

        def to_tensor_transform(self, sample: Any) -> Tuple[Tensor, Tensor]:
            x, y = sample
            x = torch.from_numpy(x).float()
            y = torch.tensor(y, dtype=torch.float)
            return x, y

        def predict_load_data(self, data: ND) -> ND:
            return data

        def predict_to_tensor_transform(self, sample: ND) -> ND:
            return torch.from_numpy(sample).float()

4. Fitting
----------

> For this task, we will be using ``scikit-learn`` `Diabetes

Reviewer 2: Not for the task for the example.

Reviewer 2: Tasks should be independent of data

Author: Done !

> testing) or override ``training_step``, ``validation_step``, and

Reviewer: can we use :func:`.... for all of these?

``test_step`` individually. These methods behave identically to PyTorch
Lightning’s
`methods <https://pytorch-lightning.readthedocs.io/en/latest/common/lightning_module.html#methods>`__.

> Here is the pseudo code behind :class:`~flash.core.model.Task` step.

Reviewer: Maybe a custom step will be more useful?

Example::

    def step(self, batch: Any, batch_idx: int) -> Any:
        """
        The training/validation/test step. Override for custom behavior.
        """
        x, y = batch
        y_hat = self(x)
        # compute the logs, loss and metrics as an output dictionary
        return output

3.a The DataModule API
----------------------

> First, let's first design the user-facing API.

Reviewer: Start with high-level. Something like:

Now that we have defined the task, we need to load our data. We will define a custom Flash :class:...DataModule to load the data from numpy arrays, and then define the preprocessing steps in a custom Flash :class:....preprocess.

a. The DataModule
Flash uses DataModules to load the datasets. Since our dataset is stored as a.... we will need to create a custom datamodule that will take that as input. We are going to create a NumpyDataModule....

b...

        ):

            preprocess = preprocess_cls()

            x_train, x_test, y_train, y_test = train_test_split(
                x, y, test_size=.20, random_state=0)

            # Make sure to call ``from_load_data_inputs``.
            dm = cls.from_load_data_inputs(
                train_load_data_input=(x_train, y_train),
                test_load_data_input=(x_test, y_test),
                preprocess=preprocess,  # DON'T FORGET TO PROVIDE THE PREPROCESS
                batch_size=batch_size,
                num_workers=num_workers
            )
            # Some metatada can be accessed from ``train_ds`` directly.
            dm.num_inputs = dm.train_dataset.num_inputs
            return dm

.. note::

    The :class:`~flash.data.data_module.DataModule` provides a ``from_load_data_inputs`` helper function. This function will take care
    of connecting the provided :class:`~flash.data.process.Preprocess` with the :class:`~flash.data.data_module.DataModule`.
    Make sure to instantiate your :class:`~flash.data.data_module.DataModule` with this helper if you rely on :class:`~flash.data.process.Preprocess`
    objects.

3.b The Preprocess API
----------------------

> .. note::

Reviewer: Let's not start with notes since it doesn't look good (esp not 2 notes).  I would take is out of the note.

3.b ....
   A :class:`~flash.data.process.Preprocess` object provides a series of hooks that can be overridden with custom data processing logic.
    It allows the user much more granular control over their data processing flow.

    The :class:`~flash.data.process.Preprocess` object reduces the engineering overhead needed for inference on raw data or
    to deploy the model in a production environment, compared to traditional
    `Dataset <https://pytorch.org/docs/stable/data.html#torch.utils.data.Dataset>`_.

and then add the note

Author: Done !

    As new concepts are being introduced, we strongly encourage the reader to click on :class:`~flash.data.process.Preprocess`
    before going further with the tutorial.

.. note::

    Why introducing :class:`~flash.data.process.Preprocess` ?

    A :class:`~flash.data.process.Preprocess` object provides a series of hooks that can be overridden with custom data processing logic.
    The user has much more granular control over their data processing flow.

    The :class:`~flash.data.process.Preprocess` object reduces the engineering overhead to make inference on raw data or
    to deploy the model in production environnement compared to traditional
    `Dataset <https://pytorch.org/docs/stable/data.html#torch.utils.data.Dataset>`_.

> Example::

Reviewer: Explain in a sentence what are the hooks you need to override in this case and what logic did you add.

    import torch
    from torch import Tensor
    import numpy as np

    ND = np.ndarray

    class NumpyPreprocess(Preprocess):

        def load_data(self, data: Tuple[ND, ND], dataset: AutoDataset) -> List[Tuple[ND, float]]:
            if self.training:
                dataset.num_inputs = data[0].shape[1]
            return [(x, y) for x, y in zip(*data)]

        def to_tensor_transform(self, sample: Any) -> Tuple[Tensor, Tensor]:
            x, y = sample
            x = torch.from_numpy(x).float()
            y = torch.tensor(y, dtype=torch.float)
            return x, y

        def predict_load_data(self, data: ND) -> ND:
            return data

        def predict_to_tensor_transform(self, sample: ND) -> ND:
            return torch.from_numpy(sample).float()

> 4. Fitting

Reviewer: To me this is no longer a part of the guide, since this is true for out-of-the-box tasks as well.

I think this should be:

1. Imports
2. Create a custom task (explain what needs to be overridden)
3. Create a custom datamodule
4. create a preprocess class
5. Optional (anything else you may need to customize?)

and then add something like:
You now have a new customized Flash Task! As for any flash task, you can now use it for fitting on a dataset and creating predictions.

add code snippet...

        [-0.0128, -0.0446, -0.0235, -0.0401, -0.0167,  0.0046, -0.0176, -0.0026, -0.0385, -0.0384],
        [-0.0237, -0.0446,  0.0455,  0.0907, -0.0181, -0.0354,  0.0707, -0.0395, -0.0345, -0.0094]])

    predictions = model.predict(predict_data)
    print(predictions)
    #out: [tensor([14.7190]), tensor([14.7100]), tensor([14.7288]), tensor([14.6685]), tensor([14.6687])]

> To customize the postprocessing of this task, you can create a :class:`~flash.data.process.Postprocess` objects and assign it to your model as follow:

Author: Done !

        [-0.0128, -0.0446, -0.0235, -0.0401, -0.0167,  0.0046, -0.0176, -0.0026, -0.0385, -0.0384],
        [-0.0237, -0.0446,  0.0455,  0.0907, -0.0181, -0.0354,  0.0707, -0.0395, -0.0345, -0.0094]])

    predictions = model.predict(predict_data)
    print(predictions)
    #out: [tensor([14.7190]), tensor([14.7100]), tensor([14.7288]), tensor([14.6685]), tensor([14.6687])]

6. Customize PostProcess
------------------------

> To customize the postprocessing of this task, you can create a :class:`~flash.data.process.Postprocess` objects and assign it to your model as follows:

Reviewer 3: ```suggestion
To customize the postprocessing of this task, you can create a :class:`~flash.data.process.Postprocess` object and assign it to your model as follows:
```

Author: Done !
